controllers/keda/scaledobject_controller.go

```diff
@@ -360,7 +360,12 @@ func (r *ScaledObjectReconciler) checkReplicaCountBoundsAreValid(scaledObject *k
 
 // ensureHPAForScaledObjectExists ensures that in cluster exist up-to-date HPA for specified ScaledObject, returns true if a new HPA was created
 func (r *ScaledObjectReconciler) ensureHPAForScaledObjectExists(ctx context.Context, logger logr.Logger, scaledObject *kedav1alpha1.ScaledObject, gvkr *kedav1alpha1.GroupVersionKindResource) (bool, error) {
-	hpaName := getHPAName(scaledObject)
+	var hpaName string
```

Comment:
I know this is an edge case, but what would happen if?
1. I do `kubectl apply` of SO with `hpaName=foo`
2. SO & HPA gets correctly created
3. KEDA Operator is scaled down for some reason
4. I do `kubectl apply` of SO with `hpaName=bar`
5. KEDA Operator is again running and will try to reconcile on the updated SO in the cluster

Reply:
I think that since the hpaName is saved in the SO status, when KEDA operator running again, it'll behave exactly the same:  
- Search for HPA with the name from SO status
- if not equal to the new hpa name -> delete the old hpa

Do you have any reason to believe that the SO.status is override while the KEDA operator is down?  

If you think it's necessary, I can try produce this steps you mentioned later when I'll be able to work on it

Reply:
So I had a chance to run it and I think I have bigger issue in this section (not just for the use case you described, but even on normal behaviour)  

I saw that even though the bottom line is that the hpa is created as expected, there are many errors along the way when trying to update the old hpa for no reason.

This is the output of step 5 of your use case, but now I see that this probably will happen on any change even without down scale the operator  

```
{"level":"info","ts":1653560500.109992,"logger":"controller.clustertriggerauthentication","msg":"Starting workers","reconciler group":"keda.sh","reconciler kind":"ClusterTriggerAuthentication","worker count":1}
{"level":"error","ts":1653560600.9423842,"logger":"controller.scaledobject","msg":"Failed to update HPA","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default","HPA.Namespace":"default","HPA.Name":"foo","error":"horizontalpodautoscalers.autoscaling \"bar\" not found","stacktrace":"github.com/kedacore/keda/v2/controllers/keda.(*ScaledObjectReconciler).ensureHPAForScaledObjectExists\n\t/Users/aviadl/Code/keda/controllers/keda/scaledobject_controller.go:390\ngithub.com/kedacore/keda/v2/controllers/keda.(*ScaledObjectReconciler).reconcileScaledObject\n\t/Users/aviadl/Code/keda/controllers/keda/scaledobject_controller.go:231\ngithub.com/kedacore/keda/v2/controllers/keda.(*ScaledObjectReconciler).Reconcile\n\t/Users/aviadl/Code/keda/controllers/keda/scaledobject_controller.go:182\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:114\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:311\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"error","ts":1653560603.795263,"logger":"controller.scaledobject","msg":"Failed to check HPA for possible update","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default","error":"horizontalpodautoscalers.autoscaling \"bar\" not found","stacktrace":"github.com/kedacore/keda/v2/controllers/keda.(*ScaledObjectReconciler).reconcileScaledObject\n\t/Users/aviadl/Code/keda/controllers/keda/scaledobject_controller.go:231\ngithub.com/kedacore/keda/v2/controllers/keda.(*ScaledObjectReconciler).Reconcile\n\t/Users/aviadl/Code/keda/controllers/keda/scaledobject_controller.go:182\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:114\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:311\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"error","ts":1653560612.2319481,"logger":"controller.scaledobject","msg":"Failed to ensure HPA is correctly created for ScaledObject","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default","error":"horizontalpodautoscalers.autoscaling \"bar\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:114\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:311\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"error","ts":1653560622.908048,"logger":"controller.scaledobject","msg":"Reconciler error","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default","error":"horizontalpodautoscalers.autoscaling \"bar\" not found","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:266\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/Users/aviadl/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.11.2/pkg/internal/controller/controller.go:227"}
{"level":"info","ts":1653560624.291129,"logger":"controller.scaledobject","msg":"Reconciling ScaledObject","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default"}
{"level":"info","ts":1653560624.2912672,"logger":"controller.scaledobject","msg":"Creating a new HPA","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default","HPA.Namespace":"default","HPA.Name":"bar"}
{"level":"info","ts":1653560624.3321779,"logger":"controller.scaledobject","msg":"Initializing Scaling logic according to ScaledObject Specification","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default"}
{"level":"info","ts":1653560624.341697,"logger":"controller.scaledobject","msg":"Reconciling ScaledObject","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default"}
{"level":"info","ts":1653560662.044199,"logger":"controller.scaledobject","msg":"Reconciling ScaledObject","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"example-scaledobject1","namespace":"default"}

```

I'll try take a further look later.  
Initial thought is that I need to create the new hpa here https://github.com/kedacore/keda/pull/3068/files#diff-191f56932c20d9d4cd383f1d52fd6091f4f734988623d4b345e75b2cf9fde7d3R166   
What do you think?

Reply:
Yeah, I think that we should chose a different approach in the implementation. I have had a hard times following the flow during the review. And I have concerns about the behavior, thus my original question.

Reply:
In the meantime I did another review, see my comment: https://github.com/kedacore/keda/pull/3068#discussion_r882540907